Foundation/NSDecimal.swift

```diff
@@ -56,74 +187,286 @@ extension Decimal {
 }
 
 extension Decimal : Hashable, Comparable {
-    public var hashValue: Int { NSUnimplemented() }
-    public static func ==(lhs: Decimal, rhs: Decimal) -> Bool { NSUnimplemented() }
-    public static func <(lhs: Decimal, rhs: Decimal) -> Bool { NSUnimplemented() }
+    internal var doubleValue: Double {
+        var d = 0.0
+        if _length == 0 && _isNegative == 0 {
+            return Double.nan
+        }
+        for i in 0..<8 {
+            let index = 8 - i - 1
+            switch index {
+            case 0:
+                d = d * 65536 + Double(_mantissa.0)
+                break
+            case 1:
+                d = d * 65536 + Double(_mantissa.1)
+                break
+            case 2:
+                d = d * 65536 + Double(_mantissa.2)
+                break
+            case 3:
+                d = d * 65536 + Double(_mantissa.3)
+                break
+            case 4:
+                d = d * 65536 + Double(_mantissa.4)
+                break
+            case 5:
+                d = d * 65536 + Double(_mantissa.5)
+                break
+            case 6:
+                d = d * 65536 + Double(_mantissa.6)
+                break
+            case 7:
+                d = d * 65536 + Double(_mantissa.7)
+                break
+            default:
+                fatalError("conversion overflow")
+            }
+        }
+        if _exponent < 0 {
+            for _ in _exponent..<0 {
+                d /= 10.0
+            }
+        } else {
+            for _ in 0..<_exponent {
+                d *= 10.0
+            }
+        }
+        return _isNegative != 0 ? -d : d
+    }
+    public var hashValue: Int {
+        return Int(bitPattern: __CFHashDouble(doubleValue))
+    }
+    public static func ==(lhs: Decimal, rhs: Decimal) -> Bool {
+        if lhs.isNaN || rhs.isNaN {
+            return false
+        }
+        if lhs.__exponent == rhs.__exponent && lhs.__lengthAndFlags == rhs.__lengthAndFlags && lhs.__reserved == rhs.__reserved {
+            if lhs._mantissa.0 == rhs._mantissa.0 &&
+                lhs._mantissa.1 == rhs._mantissa.1 &&
+                lhs._mantissa.2 == rhs._mantissa.2 &&
+                lhs._mantissa.3 == rhs._mantissa.3 &&
+                lhs._mantissa.4 == rhs._mantissa.4 &&
+                lhs._mantissa.5 == rhs._mantissa.5 &&
+                lhs._mantissa.6 == rhs._mantissa.6 &&
+                lhs._mantissa.7 == rhs._mantissa.7 {
+                return true
+            }
+        }
+        var lhsVal = lhs
+        var rhsVal = rhs
+        return NSDecimalCompare(&lhsVal, &rhsVal) == .orderedSame
```

Comment:
Sorry, I don't understand the comment on its own. Is it related to other comments or code paths?